brave/src/main/java/brave/secondary_sampling/SecondarySampling.java

```diff
@@ -154,109 +157,78 @@ public SecondarySampling build() {
   }
 
   @Override public TraceContext decorate(TraceContext context) {
-    TraceContext result = delegate.decorate(context);
-    return EXTRA_FACTORY.decorate(result);
+    TraceContext result = delegateFactory.decorate(context);
+    return SecondarySamplingDecisions.FACTORY.decorate(result);
   }
 
-  @Override public void customize(Tracing.Builder builder) {
-    builder.addFinishedSpanHandler(new SecondarySamplingSpanHandler(tagName))
-      .propagationFactory(this)
-      .alwaysReportSpans();
+  @Override public Propagation<String> get() {
+    return this;
   }
 
-  @Override public void customize(HttpTracing.Builder builder) {
-    if (httpServerSampler != null) builder.serverSampler(httpServerSampler);
+  @Override public <K> Propagation<K> create(KeyFactory<K> keyFactory) {
+    return StringPropagationAdapter.create(this, keyFactory);
   }
 
-  @Override public void customize(RpcTracing.Builder builder) {
-    if (rpcServerSampler != null) builder.serverSampler(rpcServerSampler);
+  @Override public <R> Injector<R> injector(Setter<R, String> setter) {
+    if (setter == null) throw new NullPointerException("setter == null");
+    return new SecondarySamplingInjector<>(this, setter);
   }
 
-  static final class ExtraFactory
-    extends PropagationFieldsFactory<SecondarySamplingState, Boolean, Extra> {
-    @Override public Class<Extra> type() {
-      return Extra.class;
-    }
-
-    @Override protected Extra create() {
-      return new Extra();
-    }
-
-    @Override protected Extra create(Extra parent) {
-      return new Extra(parent);
-    }
-
-    Extra create(Map<SecondarySamplingState, Boolean> initial) {
-      return initial.isEmpty() ? new Extra() : new Extra(initial);
-    }
+  @Override public <R> Extractor<R> extractor(Getter<R, String> getter) {
+    if (getter == null) throw new NullPointerException("getter == null");
+    return new SecondarySamplingExtractor<>(this, getter);
   }
 
-  static final class Extra extends MapPropagationFields<SecondarySamplingState, Boolean> {
-    Extra() {
-    }
-
-    // avoids copy-on-write for each keys on initial construction
-    Extra(Map<SecondarySamplingState, Boolean> initial) {
-      super(initial);
+  @Override public void customize(Tracing.Builder builder) {
+    builder.propagationFactory(this);
+    Set<SpanHandler> spanHandlers = new LinkedHashSet<>(builder.spanHandlers());
+    builder.clearSpanHandlers();
+    builder.addSpanHandler(new SecondarySamplingSpanHandler(tagName));
+    ZipkinSpanHandler zipkinSpanHandler = null;
+    for (SpanHandler spanHandler : spanHandlers) {
+      if (spanHandler instanceof ZipkinSpanHandler) {
+        zipkinSpanHandler = (ZipkinSpanHandler) spanHandler;
+      } else {
+        builder.addSpanHandler(spanHandler);
+      }
     }
-
-    Extra(Extra parent) {
-      super(parent);
+    if (zipkinSpanHandler != null) {
+      builder.addSpanHandler(zipkinSpanHandler.toBuilder().alwaysReportSpans(true).build());
```

Comment:
note: this is optional.. meaning if Zipkin *was* in use, we convert it to always report. Another 100% backend could be in use and we won't interfere